[CDAP-16850] Revamp schema editor

ajainarayanan opened the pull request:
**Context**
- This is a complete rewrite of schema editor.
- The goal of the schema to be more scalable and accommodate for editing attributes of certain avro schema types.

**Design**
- The schema editor design can be split into three parts,
  - Presentation - Editor
  - State management - Schema Manager
  - Validator - Schema Validator to validate schema
- Editor represents the visual part of the schema editor. At the core, it renders a list + manages changes to the schema (crud operation)
- Schema Manager is the underlying state management for any operation happening in the Editor. Internally it maintains a tree and a flat list each representing the same avro schema. Tree - for faster manipulation and Flat list - for easier rendering 
- Validator is the module used to validate the schema after each operation. We use customized version of [avsc library](https://github.com/cdapio/ui-schema-parser) to suit our needs

**Implementation**

_Schema Manager_ 
- For any avro schema generates a tree and a flat array
- For each operation (add/remove/update) the Schema Manager mutates the tree. This results in either adding a new subtree, removing a new subtree or modifying the existing subtree.
- This subtree is then flattened and inserted at appropriate position to the existing flat array.
- Tree is a direct tree representation of the avro schema
- List is a DFS traversed list generated from the above tree.

_Editor_
- List contains each node in the avro schema
- Each node in the list is identified by a unique id
- Each node contains a reference to #of ancestors
- Each row is either a two column (unions or maps) or three column like record field with name(or label), type and action buttons

**Other component changes**
- Modified `VirtualScroll` to include a `childrenUnderFoldScroll`. This is to increase the rendered child count on scroll to avoid showing empty white space. This is a common occurrence when using a virtual scroll as browser cannot catch up with scroll and rendering the underlying component quick enough.
- Modified `Select` dropdown to have a dense option to reduce padding.
- Modified `TextboxOnValium` to accept a ref prop to be able to pass on the underlying input element to the parent.
- Modified `If` to accept an additional `invisible` property to set `visibility: hidden` to children. This is to render empty space in place where the component needs to be showed.

**Things coming soon**
- Unit and Integration test for utilities, state management and editor components. - https://github.com/cdapio/cdap/pull/12455
- Replacing schema editor in the plugins modal https://github.com/cdapio/cdap/pull/12478
- Code cleanup https://github.com/cdapio/cdap/pull/12482

ajainarayanan committed:
[CDAP-16850] Adds new Schema Editor. It is a complete rewrite from scratch.

**Context**
- This is a complete rewrite of schema editor.
- The goal of the schema to be more scalable and accommodate for editing attributes of certain avro schema types.

**Design**
- The schema editor design can be split into three parts,
  - Presentation - Editor
  - State management - Schema Manager
  - Validator - Schema Validator to validate schema
- Editor represents the visual part of the schema editor. At the core, it renders a list + manages changes to the schema (crud operation)
- Schema Manager is the underlying state management for any operation happening in the Editor. Internally it maintains a tree and a flat list each representing the same avro schema. Tree - for faster manipulation and Flat list - for easier rendering
- Validator is the module used to validate the schema after each operation. We use customized version of [avsc library](https://github.com/cdapio/ui-schema-parser) to suit our needs

**Implementation**

_Schema Manager_
- For any avro schema generates a tree and a flat array
- For each operation (add/remove/update) the Schema Manager mutates the tree. This results in either adding a new subtree, removing a new subtree or modifying the existing subtree.
- This subtree is then flattened and inserted at appropriate position to the existing flat array.
- Tree is a direct tree representation of the avro schema
- List is a DFS traversed list generated from the above tree.

_Editor_
- List contains each node in the avro schema
- Each node in the list is identified by a unique id
- Each node contains a reference to #of ancestors
- Each row is either a two column (unions or maps) or three column like record field with name(or label), type and action buttons

**Other component changes**
- Modified `VirtualScroll` to include a `childrenUnderFoldScroll`. This is to increase the rendered child count on scroll to avoid showing empty white space. This is a common occurrence when using a virtual scroll as browser cannot catch up with scroll and rendering the underlying component quick enough.
- Modified `Select` dropdown to have a dense option to reduce padding.
- Modified `TextboxOnValium` to accept a ref prop to be able to pass on the underlying input element to the parent.
- Modified `If` to accept an additional `invisible` property to set `visibility: hidden` to children. This is to render empty space in place where the component needs to be showed.

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/FormInputs/Select/index.tsx:
```diff
@@ -21,35 +21,59 @@ import { IWidgetProps } from 'components/AbstractWidget';
 import { WIDGET_PROPTYPES } from 'components/AbstractWidget/constants';
 import withStyles from '@material-ui/core/styles/withStyles';
-
+import { blue } from 'components/ThemeWrapper/colors';
```
instead of importing the color, does it work if the function being passed to withStyles have the theme object?

ajainarayanan replied:
Technically we should be able to use Theme since at some point we wrap this in `ThemeWrapper` for the theme to be passed down the context. But our definition of colors in the theme is not right. `Palette` in material-ui is not defined to have colors the way we define.

Since I am still using the color used in the theme file, and even if we decide to change the color this directly gets affected. For now I would suggest let us have this. We need to fix the way we define colors in material-ui so that we could use the color we define directly from the theme without any hacks. Right now if I have to use the blue from the theme I have to type alias it to any (`outline: '1px solid ${(theme.palette as any).blue[100]}'`)

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/ArrayType/index.tsx:
```diff
@@ -0,0 +1,78 @@
+  const inputEle = React.useRef(null);
+  React.useEffect(() => {
+    if (autoFocus) {
+      if (inputEle.current) {
```
combine the two if statements?

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/ArrayType/index.tsx:
```diff
@@ -0,0 +1,78 @@
+  const [fieldTypeProperties, setFieldTypeProperties] = React.useState(typeProperties || {});
+
+  const onTypePropertiesChangeHandler = (property, value) => {
+    if (property === 'typeProperties') {
```
should we extract the types of properties that can be updated into an Enum?

ajainarayanan replied:
Yeah should have done that from the beginning. I am extracting the strings like type names and will do this as well in a separate PR. It looks like I repeat strings in a lot of places that can be an enum.

jennac3 commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaGenerator.ts:
```diff
@@ -0,0 +1,240 @@
+
+function generateArrayType(children: IOrderedChildren, nullable: boolean) {
+  const finalType = {
+    type: 'array',
```
I've noticed some types are in string form, should we define them in `enum` instead?

ajainarayanan replied:
This is a valid point. I am making this in a separate PR. Will address this in that.

jennac3 commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaGenerator.ts:
```diff
@@ -0,0 +1,240 @@
+    for (const childId of children.order) {
+      const currentChild = children[childId];
+      const { typeProperties } = currentChild;
+      if (typeProperties.symbol && typeProperties.symbol !== '') {
```
nit: I believe `if (typeProperties.symbol && typeProperties.symbol.length > 0)` may be less confusing, but this may be my biased opinion.

ajainarayanan replied:
Stating `typeProperties.symbol !== ''` is explicit. Checking for length could also mean it is an array.

jennac3 commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaGenerator.ts:
```diff
@@ -0,0 +1,240 @@
+    for (const childId of children.order) {
+      const currentChild = children[childId];
+      const { name, type, nullable: isFiledNullable } = currentChild;
+      if (!name || name === '') {
```
I believe `if (!name)` will suffice since this will check for empty string as well.

jennac3 commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaGenerator.ts:
```diff
@@ -0,0 +1,240 @@
+      const currentField = schemaTree.children[id];
+      const { name, type, nullable } = currentField;
+      // Skip the newly added rows.
+      if (!name || name === '') {
```
I believe if (!name) will suffice since this will check for empty string as well.

jennac3 commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/FieldWrapper/index.tsx:
```diff
@@ -0,0 +1,144 @@
+   */
+  const spacing = ancestors.length * INDENTATION_SPACING;
+  const firstColumn = '20px';
+  const thirdColumn = `96px`;
```
this can be just '96px'

jennac3 commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/SampleTypeFieldChecker.ts:
```diff
@@ -0,0 +1,209 @@
+    fieldsInRecords = (map1.keys as IRecordField).fields;
+  }
+  // tslint:disable-next-line: no-console
+  console.log(map1, fieldsInRecords);
```
nit: are these console.log necessary?

ajainarayanan replied:
This file is not being used anywhere.. This is to playaround with schema and typescript types. We can safely ignore this file.

jennac3 commented on cdap-ui/app/cdap/components/If/index.tsx:
```diff
@@ -14,15 +14,34 @@
+}) => {
   if (!condition) {
-    return null;
+    if (!invisible) {
```
this if (!invisible) can be moved outside so that it can be less nested.

ajainarayanan replied:
Hhmm if we flatten it out we ideally check for `invisible` twice but I agree a flat `if` check is easier to read. Have changed it.

jennac3 commented on cdap-ui/app/cdap/services/helpers.js:
```diff
@@ -490,6 +490,16 @@ function connectWithStore(store, WrappedComponent, ...args) {
 }
 
+function dumbestClone(jsonObj) {
+  let result;
```
can we just return result in `try` block?

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaManagerUtilities.ts:
```diff
@@ -0,0 +1,125 @@
+      return parseMapType(defaultMapType);
+    case 'record': {
+      return parseComplexType(defaultRecordType);
+    }
```
nit: unnecessary `{}`

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaParser.ts:
```diff
@@ -0,0 +1,431 @@
+    case 'record': {
+      record = parseRecordType(type);
+      break;
+    }
```
nit: remove `{}`

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaParser.ts:
```diff
@@ -0,0 +1,431 @@
+  });
+  const mapValuesSubType = getMapSubType(valuesType, {
+    simpleType: InternalTypesEnum.MAP_VALUES_SIMPLE_TYPE,
+    complexType: InternalTypesEnum.MAP_VALUES_COMPLEX_TYPE_ROOT,
```
do these need to be exposed here? can these value just be inside the `getMapSubType`?

ajainarayanan replied:
I am using it for both keys and values type. Depending on the usage the simpleType and complexType differs. Do you recommend anything specific we can do here?

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/RowButtons/FieldAttributes/RecordEnumAttributes.tsx:
```diff
@@ -0,0 +1,89 @@
+      margin: '10px 0',
+    },
+  },
+});
```
this style seemed to be the same across a few places. Does it make sense to extract?

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/SchemaHelpers.ts:
```diff
@@ -0,0 +1,190 @@
+    type = complexType.filter((t) => t !== 'null').pop();
+    type = type.type;
+  } else {
+    type = c.type;
```
minor optimization, should we just cloneDeep here instead of at the beginning of the function? It seems like only this part requires the cloned type object

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/index.tsx:
```diff
@@ -0,0 +1,125 @@
+    };
+  }
+
+  public componentWillReceiveProps(nextProps) {
```
I think for this particular case, we can use the `getDerivedStateFromProps` lifecycle and memoizing it since you need to update the state based on props change. Might future-proof this since componentWillReceiveProps is being deprecated

ajainarayanan replied:
In a future PR I am getting rid of this function. Any component that needs to update the schemaeditor with new schema has to reload the component. It became a hassle when integrating with the plugins.

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/index.tsx:
```diff
@@ -0,0 +1,125 @@
+  SchemaValidatorConsumer,
+  SchemaValidatorProvider,
+} from 'components/AbstractWidget/SchemaEditor/SchemaValidator';
+import { dumbestClone as cloneDeep } from 'services/helpers';
```
can we use this just as dumbestClone and not alias it with cloneDeep? Might be easier to detect if we need to refactor later simply through search

ajainarayanan replied:
When renaming the function for some reason typescript created the alias here instead of the helpers. Have changed.

yukiej commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaGenerator.ts:
```diff
@@ -0,0 +1,240 @@
+  if (Array.isArray(children.order)) {
+    for (const childId of children.order) {
+      const currentChild = children[childId];
+      const { name, type, nullable: isFiledNullable } = currentChild;
```
I don't see "isFiledNullable" defined anywhere - should this just be "nullable"? (same for line 145)

njbriggs replied:
Should this be `isFieldNullable`?

yukiej commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaManagerUtilities.ts:
```diff
@@ -0,0 +1,125 @@
+import isEmpty from 'lodash/isEmpty';
+
+function getInternalType(tree: INode) {
+  const hasChildren = tree.children ? Object.keys(tree.children).length : 0;
```
nit: The naming of this (currently numeric) variable sounds like it is a boolean, and it looks like it is being used as a boolean value. Maybe it can be modified slightly to have a boolean value.

yukiej commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/RowButtons/FieldAttributes/RecordEnumAttributes.tsx:
```diff
@@ -0,0 +1,89 @@
+          pluginProperty={{
+            name: 'doc',
+            macroSupported: false,
+            description: 'doc for the record',
```
Nit: Should this be capitalized for consistency?

ajainarayanan replied:
I actually got feedback for this even before creating the PR and yet missed this one 🙄 . Thanks for catching this!.

yukiej commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/SchemaTypes.ts:
```diff
@@ -0,0 +1,177 @@
+
+type IDisplayType = ISimpleType | IComplexTypeNames;
+
+type ISimpleTypeNullable = Array<ISimpleType | 'null'>;
```
Should this be null (without the single quotes)?

ajainarayanan replied:
No. `'null'` should be a string. `'null'` in avro is actually a valid type. We treat it as auxillary information for any type because it makes user experience much simpler.

yukiej commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/SchemaValidator/index.tsx:
```diff
@@ -0,0 +1,148 @@
+        const entireSchema = SchemaGenerator(schemaTree);
+        cdapavsc.parse(entireSchema.schema, { wrapUnions: true });
+      } catch (e) {
+        return { error: e.message, fieldIdToShowError: ancestors[0] };
```
Do we know for sure that the error is always in e.message?

ajainarayanan replied:
Yeah this is the standard javascript error thrown from within the library.

yukiej commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/SingleColumnWrapper/index.tsx:
```diff
@@ -0,0 +1,31 @@
+const SingleColumnWrapper = withStyles(() => {
+  return {
+    root: {
+      width: '132px', // 50 + 100 + 10 from the mapwrapper
```
Nit: these numbers don't add up to 132

njbriggs commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/EditorTypes.ts:
```diff
@@ -0,0 +1,104 @@
+  property?: string;
+  value?: string | ITypeProperties;
+  type:
+    | OperationTypesEnum.UPDATE
```
This has a leading `|`. I don't know if it matters though.

ajainarayanan replied:
So apparently this seems to be a valid typescript syntax. This is done by prettier's auto format.

njbriggs commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/RowButtons/AddRowButton.tsx:
```diff
@@ -0,0 +1,29 @@
+
+import * as React from 'react';
+import AddIcon from '@material-ui/icons/Add';
+import { IconWrapper } from 'components/AbstractWidget/SchemaEditor/RowButtons/RemoveRowButton';
```
Seems odd to import `IconWrapper` from `RemoveRowButton`?

njbriggs commented on cdap-ui/app/cdap/components/If/index.tsx:
```diff
@@ -14,15 +14,34 @@
+const If: React.FC<IIfComponentProps> = ({
+  condition,
+  children,
+  invisible = false,
```
Do we want to add this to `If` or create a new `Hidden` component? `If` seems like it should stay a one-purpose component.

ajainarayanan replied:
The thought process was if the condition is false the children under the `If` doesn't get rendered. We can choose to make it invisible (occupying the space) or set it to display none there by avoiding browser to paint anything. 

If you feel we should have this as a separate component I can make it but it will still do the same thing done by `If` except set the visibility to hidden instead of display none.

ajainarayanan committed:
Integrates new Schema Editor to pipeline plugins

**Approach**
- Compose `SchemaEditor` into a new component named `PluginSchemaEditor` along with other `Actions` available on the schema.
- Have retained the logic to enable/disable the schema actions in angular for the most part.
- Some of the logic, like export, are still in react. We use the same event based system for import/export/clear
- PluginSchemaEditor will also determine the height of the container to determine the number of rows that needs to be rendered (required by the virtual scroll)
- Changes avro schema types from string to enum
- Fixes `SchemaManager` to handle logical types for all complex types
- Adds validation on import of schema
- Adds the ability to set `disabled` for the schema editor to become non-editable.

**Note**
- This change achieves feature parity with existing schema editor.
- However at any point if the user thinks they are experiencing any regressions they can switch off the experiment which will enable them to use the older schema editor
- The experiment will be on by default.
- All the changes to angular side are additive and does not remove any existing functionality.

yukiej commented on cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaManager.ts:
```diff
@@ -0,0 +1,596 @@
+    getInstance: () => schemaTreeInstance,
+  };
+}
+export { SchemaManager, INode, ISchemaManager, IOnChangeReturnType, ISchemaManagerOptions };
```
Is exporting INode here a typo? Since we imported it from SchemaParser

yukiej reviewed (approved):
One small comment, otherwise lgtm!

jennac3 commented on cdap-ui/app/cdap/components/AbstractWidget/FormInputs/Select/index.tsx:
```diff
@@ -21,35 +21,59 @@ import { IWidgetProps } from 'components/AbstractWidget';
       padding: '7px 18px 7px 12px',
       '&:focus': {
         backgroundColor: 'transparent',
+        outline: `1px solid ${blue[100]}`,
```
we can use `theme` in `withStyles` I believe.

`withStyles((theme))`

ajainarayanan replied:
https://github.com/cdapio/cdap/pull/12450#discussion_r455414457

njbriggs reviewed (approved):
LGTM

ajainarayanan committed:
Load default experiments to localStorage on initial load + sets new schema editor experiment to true by default

elfenheart commented on cdap-ui/app/cdap/components/Lab/experiment-list.tsx:
```diff
@@ -35,6 +35,6 @@ export default [
       'Demo for new SchemaEditor. Includes complete rewrite in React + perf improvements',
```
nit: perhaps remove the `Demo for` part from this description since this lab is not for the demo page anymore?

elfenheart reviewed (approved):
Tiny nit, otherwise 👍  LGTM
